src/SourceGenerators/EquatableArray.cs

```diff
@@ -0,0 +1,70 @@
+// <copyright file="EquatableArray.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System.Collections;
+
+namespace SourceGenerators;
+
+internal readonly record struct EquatableArray<T>
+    : IEnumerable<T>
+    where T : IEquatable<T>
+{
+    private readonly T[] _array;
+
+    public EquatableArray(T[] array)
+    {
+        _array = array;
+    }
+
+    public int Count => _array.Length;
+
+    public bool Equals(EquatableArray<T> array)
+    {
+        return _array.AsSpan().SequenceEqual(array.AsSpan());
```

Comment:
I don't think this matter for the specific case at hand, but, seems a good thing for the type in general.

```suggestion
        return _array == array || _array.AsSpan().SequenceEqual(array.AsSpan());
```